### DISCUSSION TITLE
Trying to subclass `eglot-lsp-server`

### DISCUSSION THREAD

**Discussion by benbellick** [joaotavora/eglot#1423] [Q&A] [answered]
Hi! I actually had been trying to implement #1420 myself by doing what the documentation had suggested, which is to alter the `eglot-server-programs` so that the value for c++-mode goes from this:
```elisp
                                ((c-mode c-ts-mode c++-mode c++-ts-mode objc-mode)
                                 . ,(eglot-alternatives
                                     '("clangd" "ccls")))
```
to this:
```elisp
                                ((c-mode c-ts-mode c++-mode c++-ts-mode objc-mode)
                                 . ('eglot-lsp-server . ("clangd"))))
```

Then, I would later replace `'eglot-lsp-server` with my own custom subclass. This was informed by the documentation string for the variable which says contact can be (emphasis mine)

>   A cons (CLASS-NAME . INITARGS) where CLASS-NAME is a symbol
>   designating a subclass of `eglot-lsp-server`, for representing
>   experimental LSP servers.  INITARGS is a keyword-value plist
>   used to initialize the object of CLASS-NAME, or _**a plain list
>   interpreted as the previous descriptions of CONTACT.**_  In the
>   latter case that plain list is used to produce a plist with a
>   suitable :PROCESS initarg to CLASS-NAME.  The class
>   `eglot-lsp-server` descends from `jsonrpc-process-connection`,
>   which you should see for the semantics of the mandatory
>   :PROCESS argument.

However I get the error
> Wrong type argument: processp, nil

Any suggestions on what I am doing wrong here? Am I misunderstanding the docs? Anyways, thanks! this package rocks

**Comment by joaotavora** [comment]
This would seem to be a bug. Either in the doc or in the implementation. I have to check.

**Reply by joaotavora** [reply]
> This would seem to be a bug. Either in the doc or in the implementation. I have to check.

Never mind, no it isn't.

**Comment by joaotavora** [comment, marked as answer]
> Any suggestions on what I am doing wrong here? 

You have to brush up on your Lisp quoting rules (and I must play closer attention):

```
(setq-local eglot-server-programs
     '((c-mode c-ts-mode c++-mode c++-ts-mode objc-mode)
                                 . (eglot-lsp-server . ("clangd"))))
```

Is how it should be.  This of course is the same as 

```
(setq-local eglot-server-programs
     '((c-mode c-ts-mode c++-mode c++-ts-mode objc-mode)  eglot-lsp-server "clangd"))
```

But what you had was wrong (for Eglot's purposes) because `'` means the same as applying the form `quote`.  So this:

````
(setq-local eglot-server-programs
     '((c-mode c-ts-mode c++-mode c++-ts-mode objc-mode)
                                 . ('eglot-lsp-server . ("clangd"))))
````

is the same as this (I think):

```
(setq-local eglot-server-programs
     (quote ((c-mode c-ts-mode c++-mode c++-ts-mode objc-mode)
                                 . ((quote eglot-lsp-server) . ("clangd")))))
```

Which is wrong: has too much quoting.

**Reply by benbellick** [reply]
Nice catch, thank you!!
